vhost-device-scmi/src/devices/iio.rs

```diff
@@ -347,7 +347,12 @@ impl IIOSensor {
 
     fn set_sensor_name_from_file(&mut self, path: &PathBuf) {
         match fs::read_to_string(path) {
-            Ok(name) => self.sensor_mut().name = name,
+            Ok(name) => {
+                //Do not override previous name
+                if self.sensor_mut().name.is_empty() {
+                    self.sensor_mut().name = name;
+                }
+            },
```

Comment:
Can you give a bit more info on what this is solving? The commit message is:

> 
> It can avoid that program exit unexpectedly
> when the length of native sensor name is too long.
> 

What happens when `name` is too long and it exits?

Also, it seems like you can skip the file IO here if name is not empty:

```
    fn set_sensor_name_from_file(&mut self, path: &PathBuf) {
        // Do not override previous name
        if !self.sensor_mut().name.is_empty() {
            return;
        }
        match fs::read_to_string(path) {
            Ok(name) => {
                self.sensor_mut().name = name;
            },
```

Reply:
If a native iio device has it own name and too long, it will cause this error.

For example, 
We started vhost-user-scmi device like:

/usr/bin/vhost-device-scmi --socket-path /tmp/vhost-user/virtio-vm3-vscmi \
        -d iio,path=/sys/bus/iio/devices/iio:device1,channel=in_accel,name=scmi.iio.accel 

And we expected its name as "scmi.iio.accel" in guest.

But if backend can get sensor name from "/sys/bus/iio/devices/iio:device0/name", it will forcely override the name we set by command (e.g. asm330lhh_accel)

In ARM SCMI spec, the command "SENSOR_DESCRIPTION_GET" should return sensor descriptor.
And sensor_name introduction is here
	uint8 sensor_name[16]
	A NULL terminated UTF-8 format string with the sensor name, of up to 16 bytes.

Then in file "vhost-device-scmi/src/scmi.rs", the assert "v_len < size" in function "from" will fail.

We think, once backend sets sensor name in command line, it will not be overrided

Reply:
That makes sense. I think it might be better to check if name is not empty in `register_iio_file` when file name is `name` then, and leave `set_sensor_name_from_file` unaltered

I don't see any particular requirements in the spec about sensor names other than being 15 bytes plus the null terminator (unless it's an extended sensor name). Plus a String is allowed to have null bytes in it so it should really be a Option<CString> with a truncation...

> Then in file "vhost-device-scmi/src/scmi.rs", the assert "v_len < size" in function "from" will fail.

Maybe we should remove the assert as well for now and just ignore the extra bytes. What do you think?

Reply:
Your suggestion is great, maybe motificiation like this is better?

```
fn register_iio_file(&mut self, file: fs::DirEntry, axes: &mut Vec<Axis>) {
        let channel = self.channel.to_str().unwrap();
        let os_file_name = file.file_name();
        let file_name = os_file_name.to_str().unwrap_or_default();
        let raw_suffix = "_raw";
-       if file_name == "name" {
+       if file_name == "name" && self.sensor_mut().name.is_empty() {
```

And for this, you can refer to the document "[Arm System Control and Management Interface Platform Design Document](https://developer.arm.com/documentation/den0056/e?lang=en)" 
> I don't see any particular requirements in the spec about sensor names other than being 15 bytes plus the null terminator (unless it's an extended sensor name). Plus a String is allowed to have null bytes in it so it should really be a Option<CString> with a truncation...

In **_4.7.2.5.1 Sensor Descriptor_**

It define SENSOR_DESC structure

```
uint8 sensor_name[16]

A NULL terminated UTF-8 format string with the sensor name, of up to 16 bytes. 
When Bit[29] of sensor_attributes_low field is set to 1, this field contains the lower 15 bytes of the NULL terminated sensor name.
```

According to spec, if  Bit[29] of sensor_attributes_low "_Extended sensor name_" support, sensor name can be greater than 16 bytes.

And in my opinion, ignore the extra bytes instead of assert is better than current.

Reply:
> Your suggestion is great, maybe motificiation like this is better?

Exactly, yes!

> And in my opinion, ignore the extra bytes instead of assert is better than current.

Yep, we should use asserts when it's no longer possible to continue with a correct state, failing for this reason is not very reasonable

Reply:
I will re-upload this patch, and thanks for your suggestion.

Reply:
Hello, epilys. For this modification, I have a question about the test function
```
fn test_sensor_name_from_fs() {
        let directory = IIODirectory::new(&[("in_accel_raw", "123"), ("name", "foo")]);
        let mut sensor =
            make_iio_sensor(&directory, "in_accel".to_owned(), Some("accel".to_owned()));
        sensor.initialize().unwrap();
        assert_eq!(sensor.sensor.name, "foo");
    }
```

In previous logic, the sensor's name will be override from fs. But after my patch, it will fail.

Should this test be motidified?

Reply:
And the three ways to set iio sensor name should have a prority order
In my opinion, the order list should be

1.  From parameter
2.  From fs
3.  Defalut

Reply:
Intuitively I'd say the unit test should not overwrite the name anymore but let's ask @mz-pdm who's familiar with the crate, what do you think?

Reply:
That's well

Reply:
I agree with the order list above. I can't recall any reason why the file system name should have precedence over the command line name, I probably thought only about the situation when there is no file system name.

That said, I think the three related tests (test_sensor_name_from_fs, test_sensor_name_from_params, test_default_sensor_name) should still pass. Why does test_sensor_name_from_fs fail? It should check the situation when there is a file system name but no name on the command line.

A new test should be added, similar to test_sensor_name_from_params, with `("name", "override-me")` added to IIODirectory::new. That would check that the command line name takes precedence.

Reply:
If with my modification 
```
-        if file_name == "name" && self.sensor.name.is_empty() {
+        if file_name == "name" && self.sensor.name == "iio" {
```
The previous unit test "_test_sensor_name_from_fs_" will failed, because it make sensor with name "accel" and 
"_assert_eq!(sensor.sensor.name, "foo")_", left is "accel“


**This is previous test, name fs exist, and make sensor with "accel" name, with my modification, it will not be override to "foo", then assert failed.**
```
fn test_sensor_name_from_fs() {
         let directory = IIODirectory::new(&[("in_accel_raw", "123"), ("name", "foo")]);
         let mut sensor =
            make_iio_sensor(&directory, "in_accel".to_owned(), Some("accel".to_owned()));
         sensor.initialize().unwrap();
         assert_eq!(sensor.sensor.name, "foo");
     }
```

Reply:
Ah, right. Your change to test_sensor_name_from_fs is all right.

As for test_sensor_name_from_params, I would keep both the original version and your modified version (as two separate tests). This way, the tests would cover the situations when the command line name is provided and the file system name either exists (and is overridden) or doesn't exist (and the command line name still applies without any error).

Reply:
That's well